{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Motifs in DNase I peaks\n",
    "\n",
    "```{note}\n",
    "This exercise is *optional*!\n",
    "```\n",
    "\n",
    "**Exercise 13.1**\n",
    "\n",
    "The exercise for this week is to compare the relation between transcription factor motif presence and DNase I signal in different cell types.\n",
    "\n",
    "The result of your code should be a figure like the following (the values here are not real, but only shown for illustrative purposes):\n",
    "\n",
    "![Figure 1: Example heatmap for this exercise.](img/dnase_motifs_heatmap.png)\n",
    "\n",
    "This heatmap shows, for every cell type and motif combination, the difference of the mean signal for regions with a motif compared to regions without a motif.\n",
    "\n",
    "**Data**\n",
    "\n",
    "In the directory `/content/gdrive/MyDrive/CFB_2022/dnase` you will find two files:\n",
    "\n",
    "* `DNase_table.tsv` \n",
    "* `motifs.txt`\n",
    "\n",
    "\n",
    "The file `DNase_table.tsv` looks like this:\n",
    "\n",
    "![Figure 2: Contents of DNase_table.tsv.](img/dnase_table.png)\n",
    "\n",
    "This tab-separated file contains DNase I signals for six hematopoietic cell types in 10,000 genomic regions. The DNase I reads were counted, log2 transformed and normalized by scaling.\n",
    "\n",
    "The file `motifs.txt` contains the name and consensus sequence for seven motifs.\n",
    "\n",
    "Use the shell command `wget` to download UCSC Genome Browserâ€™s hg38 human reference sequence. Note that this is a gzip archive (`.gz`) of the file `hg38.fa`. Use `gunzip` to extract `hg38.fa` from the archive:\n",
    "\n",
    "```\n",
    "!wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\n",
    "!gunzip hg38.fa.gz\n",
    "```\n",
    "\n",
    "You now have `hg38.fa` in your Google Colab `/content` folder. Recall that this folder is temporary and will be cleared in a few hours. You can always re-download the file, or you can save it in your Google Drive. However, this file is ca. 3 Gb, so make sure you have enough space on your Google Drive if you choose to save it.\n",
    "\n",
    "\n",
    "**Retrieving sequences from a genome FASTA file.**\n",
    "\n",
    "You can use the `pyfaidx` module to retrieve specific sequences from a FASTA file based on genomic coordinates. See the documentation of the module [here](https://pythonhosted.org/pyfaidx/).\n",
    "\n",
    "First, you create a FASTA object.\n",
    "\n",
    "```\n",
    "!pip install pyfaidx\n",
    "\n",
    "from pyfaidx import Fasta\n",
    "genome = Fasta(\"hg38.fa\")\n",
    "```\n",
    "\n",
    "Then, you can retrieve sequences in a dictionary-like approach.\n",
    "\n",
    "```\n",
    "my_seq = genome[\"chr2\"][5000000:5000020]\n",
    "print(my_seq.seq)\n",
    "```\n",
    "\n",
    "```\n",
    "TCATGACCATAGAGAACAGA\n",
    "```\n",
    "\n",
    "\n",
    "**Approach**\n",
    "\n",
    "Think about your approach before you start writing code!\n",
    "You will have to:\n",
    "\n",
    "* scan sequences with the motif consensus sequences;\n",
    "* determine, for every motif, which sequences have a match to the motif;\n",
    "* calculate the mean signal, depending on the motif occurrence;\n",
    "* calculate the difference in mean signal of sequences with a motif match and without a motif match.\n",
    "\n",
    "**Questions**\n",
    "\n",
    "* Make a correlation plot of the DNase signal in these hematopoietic cel types? Is this what you expect?\n",
    "* Which motifs are specific for certain cell types?\n",
    "* Does this agree with the literature?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.7 ('testing')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  },
  "vscode": {
   "interpreter": {
    "hash": "d3e95029f17eb34d2d190ec57747705a24a9150ea64dec8a02f0c1132051a0c3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}