
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>13. Motifs in DNase I peaks &#8212; NWI-BM066A Computation for Biologists</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="14. Creating a DNAseI footprint profile" href="footprints.html" />
    <link rel="prev" title="12. Visualization" href="visualization.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">NWI-BM066A Computation for Biologists</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="01_introduction.html">
                    Introduction
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="first_exercises.html">
   1. First exercises
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="lists.html">
   2. Lists
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="control_flow.html">
   3. Control flow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="loops.html">
   4. Loops
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="functions_modules.html">
   5. Functions and modules
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dictionaries.html">
   6. Dictionaries
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="extra_exercises.html">
   7. Assembling the pieces
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="files.html">
   8. Files
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="scripts.html">
   9. Scripts
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="debugging.html">
   10. Debugging
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pandas.html">
   11. Introduction to pandas DataFrames
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="visualization.html">
   12. Visualization
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   13. Motifs in DNase I peaks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="footprints.html">
   14. Creating a DNAseI footprint profile
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/YuLab-RU/CFB_book/master?urlpath=tree/content/dnase_motifs.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/YuLab-RU/CFB_book/"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/YuLab-RU/CFB_book//issues/new?title=Issue%20on%20page%20%2Fcontent/dnase_motifs.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/YuLab-RU/CFB_book/edit/master/content/dnase_motifs.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/content/dnase_motifs.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Motifs in DNase I peaks</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="motifs-in-dnase-i-peaks">
<h1><span class="section-number">13. </span>Motifs in DNase I peaks<a class="headerlink" href="#motifs-in-dnase-i-peaks" title="Permalink to this headline">#</a></h1>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This exercise is <em>optional</em>!</p>
</div>
<p><strong>Exercise 13.1</strong></p>
<p>The exercise for this week is to compare the relation between transcription factor motif presence and DNase I signal in different cell types.</p>
<p>The result of your code should be a figure like the following (the values here are not real, but only shown for illustrative purposes):</p>
<p><img alt="Figure 1: Example heatmap for this exercise." src="../_images/dnase_motifs_heatmap.png" /></p>
<p>This heatmap shows, for every cell type and motif combination, the difference of the mean signal for regions with a motif compared to regions without a motif.</p>
<p><strong>Data</strong></p>
<p>In the directory <code class="docutils literal notranslate"><span class="pre">/content/gdrive/MyDrive/CFB_2022/dnase</span></code> you will find two files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">DNase_table.tsv</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">motifs.txt</span></code></p></li>
</ul>
<p>The file <code class="docutils literal notranslate"><span class="pre">DNase_table.tsv</span></code> looks like this:</p>
<p><img alt="Figure 2: Contents of DNase_table.tsv." src="../_images/dnase_table.png" /></p>
<p>This tab-separated file contains DNase I signals for six hematopoietic cell types in 10,000 genomic regions. The DNase I reads were counted, log2 transformed and normalized by scaling.</p>
<p>The file <code class="docutils literal notranslate"><span class="pre">motifs.txt</span></code> contains the name and consensus sequence for seven motifs.</p>
<p>Use the shell command <code class="docutils literal notranslate"><span class="pre">wget</span></code> to download UCSC Genome Browser’s hg38 human reference sequence. Note that this is a gzip archive (<code class="docutils literal notranslate"><span class="pre">.gz</span></code>) of the file <code class="docutils literal notranslate"><span class="pre">hg38.fa</span></code>. Use <code class="docutils literal notranslate"><span class="pre">gunzip</span></code> to extract <code class="docutils literal notranslate"><span class="pre">hg38.fa</span></code> from the archive:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>!wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
!gunzip hg38.fa.gz
</pre></div>
</div>
<p>You now have <code class="docutils literal notranslate"><span class="pre">hg38.fa</span></code> in your Google Colab <code class="docutils literal notranslate"><span class="pre">/content</span></code> folder. Recall that this folder is temporary and will be cleared in a few hours. You can always re-download the file, or you can save it in your Google Drive. However, this file is ca. 3 Gb, so make sure you have enough space on your Google Drive if you choose to save it.</p>
<p><strong>Retrieving sequences from a genome FASTA file.</strong></p>
<p>You can use the <code class="docutils literal notranslate"><span class="pre">pyfaidx</span></code> module to retrieve specific sequences from a FASTA file based on genomic coordinates. See the documentation of the module <a class="reference external" href="https://pythonhosted.org/pyfaidx/">here</a>.</p>
<p>First, you create a FASTA object.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>!pip install pyfaidx

from pyfaidx import Fasta
genome = Fasta(&quot;hg38.fa&quot;)
</pre></div>
</div>
<p>Then, you can retrieve sequences in a dictionary-like approach.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;chr2&quot;</span><span class="p">][</span><span class="mi">5000000</span><span class="p">:</span><span class="mi">5000020</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">TCATGACCATAGAGAACAGA</span>
</pre></div>
</div>
<p><strong>Approach</strong></p>
<p>Think about your approach before you start writing code!
You will have to:</p>
<ul class="simple">
<li><p>scan sequences with the motif consensus sequences;</p></li>
<li><p>determine, for every motif, which sequences have a match to the motif;</p></li>
<li><p>calculate the mean signal, depending on the motif occurrence;</p></li>
<li><p>calculate the difference in mean signal of sequences with a motif match and without a motif match.</p></li>
</ul>
<p><strong>Questions</strong></p>
<ul class="simple">
<li><p>Make a correlation plot of the DNase signal in these hematopoietic cel types? Is this what you expect?</p></li>
<li><p>Which motifs are specific for certain cell types?</p></li>
<li><p>Does this agree with the literature?</p></li>
</ul>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./content"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="visualization.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">12. </span>Visualization</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="footprints.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">14. </span>Creating a DNAseI footprint profile</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Simon van Heeringen, Arjen Brinkman, Kees Albers, Rosemary Yu<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>